kolibri/auth/models.py

```diff
@@ -1,497 +1,428 @@
+We have three main abstractions: Users, Collections, and Roles.
 
-Collections are hierarchical groups of users. Users belong to one or more Collections, and Collections can belong to
-other Collections. Collections are subdivided into several pre-defined levels.
+Users represent people, like students in a school, teachers for a classroom, or volunteers setting up informal
+installations. There are two main user types, FacilityUser and DeviceOwner. A FacilityUser belongs to a particular
+facility, and has permissions only with respect to other data that is associated with that facility. A DeviceOwner
+is not associated with a particular facility, and has global permissions for data on the local device. FacilityUser
+accounts (like other facility data) may be synced across multiple devices, whereas a DeviceOwner account is specific
+to a single installation of Kolibri.
 
-Roles belong to collections, and represent permissions. Users have one or more Roles. For instance, Classes (a type
-of Collection) have an associated Coach Role -- that Coach has permission to view related User data for Users in the
-Class.
+Collections form a hierarchy, with Collections able to belong to other Collections. Collections are subdivided
+into several pre-defined levels (Facility > Classroom > LearnerGroup).
+
+A Role connects a FacilityUser and a Collection, for purposes of membership and permissions, along with a specified
+"kind" (such as "admin", "coach", or "learner"). Users can have more than one Role. For instance, a User may have a
+coach role for one Classroom, and an admin role for another.
 """
+
 from __future__ import absolute_import, print_function, unicode_literals
 
-from django.contrib.auth.models import AbstractBaseUser, _user_get_all_permissions, _user_has_module_perms, \
-    _user_has_perm
+from django.contrib.auth.models import AbstractBaseUser
 from django.core import validators
 from django.db import models
-from django.utils.translation import ugettext_lazy as _
+from django.db.utils import IntegrityError
 from django.utils import timezone
-from mptt.models import TreeForeignKey, MPTTModel
+from django.utils.translation import ugettext_lazy as _
+from kolibri.core.errors import KolibriValidationError
+from mptt.models import MPTTModel, TreeForeignKey
 
-from kolibri.core.errors import KolibriError
 
+class FacilityDataset(models.Model):
+    """
+    FacilityDataset stores high-level metadata and settings for a particular facility. It is also the
+    model that all models storing facility data (data that is associated with a particular facility, and that inherits
+    from AbstractFacilityDataModel) foreign key onto, to indicate that they belong to this particular facility.
+    """
 
-class KolibriValidationError(KolibriError):
-    pass
+    description = models.TextField(blank=True)
+    location = models.CharField(max_length=200, blank=True)
 
+    allow_signups = models.BooleanField(default=True)
 
-class BaseUser(AbstractBaseUser):
+
```

Comment:
@MCGallaspy The class below this comment is new, and handles the "dataset inference" pieces we discussed. As you can see in the test files, this means that the app programmer should very rarely need to pass in a dataset, as it will be inferred at time of model creation.